runtime/protocol/http-client-engines/http-client-engine-okhttp/jvm/src/aws/smithy/kotlin/runtime/http/engine/okhttp/OkHttpUtils.kt

```diff
@@ -189,3 +190,27 @@ private fun URI.toUrl(): Url {
         fragment = uri.fragment?.takeIf { it.isNotBlank() }
     }
 }
+
+internal inline fun<T> mapOkHttpExceptions(block: () -> T): T =
+    try {
+        block()
+    } catch (ex: IOException) {
+        throw HttpException(ex.message, ex, ex.errCode(), ex.isRetryable())
+    }
+
+private fun Exception.isRetryable(): Boolean = isCauseOrSuppressed<ConnectException>()
+private fun Exception.errCode(): HttpErrorCode = when {
+    isConnectTimeoutException() -> HttpErrorCode.CONNECT_TIMEOUT
+    isCauseOrSuppressed<SocketTimeoutException>() -> HttpErrorCode.SOCKET_TIMEOUT
+    isCauseOrSuppressed<SSLHandshakeException>() -> HttpErrorCode.TLS_NEGOTIATION_ERROR
+    else -> HttpErrorCode.SDK_UNKNOWN
+}
+
+private fun Exception.isConnectTimeoutException(): Boolean =
+    findCauseOrSuppressed<SocketTimeoutException>()?.message?.contains("connect", ignoreCase = true) == true
+private inline fun <reified T> Exception.isCauseOrSuppressed(): Boolean = findCauseOrSuppressed<T>() != null
+private inline fun <reified T> Exception.findCauseOrSuppressed(): T? {
+    if (this is T) return this
+    if (cause is T) return cause as T
+    return suppressedExceptions.firstOrNull { it is T } as? T
+}
```

Comment:
**Question**: Causes and suppressions are recursive. Do we need to check deeper in those hierarchies?

Reply:
I don't suspect so based on what I've seen, this can all be tweaked if we find additional edge cases of course.